## Foundations of Computer Science
This course has two aims. The first is to teach programming. The second is to
present some fundamental principles of computer science, especially algorithm
design. Most students will have some programming experience already, but there
are few people whose programming cannot be improved through greater knowledge
of basic principles.

The programming in this course is based on the language [OCaml](https://ocaml.org)
and mostly concerns the functional programming style.

**This is a work-in-progress port of Lawrence C. Paulson's 2018-19 Cambridge
course notes. Contact: Anil Madhavapeddy avsm2, Jon Ludlam jjl25**


## Lecture 1: Algorithms and Recursive Functions 

### Basic Concepts in Computer Science

- Computers: a child can use them; **nobody** can fully understand them!
- We can master complexity through levels of abstraction.
- Focus on 2 or 3 levels at most!

**Recurring issues:**
- what services to provide at each level
- how to implement them using lower-level services
- the interface: how the two levels should communicate

A basic concept in computer science is that large systems can only be
understood in levels, with each level further subdivided into functions or
services of some sort. The interface to the higher level should supply the
advertised services. Just as important, it should block access to the means by
which those services are implemented. This _abstraction barrier_ allows one
level to be changed without affecting levels above. For example, when a
manufacturer designs a faster version of a processor, it is essential that
existing programs continue to run on it. Any differences between the old and
new processors should be invisible to the program.

Modern processors have elaborate specifications, which still sometimes leave
out important details. In the old days, you then had to consult the circuit
diagrams.

### Example 1: Dates

- Abstract level: dates over a certain interval
- Concrete level: typically 6 characters: `YYMMDD` (where each character is represented by 8 bits)
- Date crises caused by __inadequate__ internal formats:
  * Digital’s PDP-10: using 12-bit dates (good for at most 11 years)
  * 2000 crisis: 48 bits could be good for lifetime of universe!

Digital Equipment Corporation’s date crisis occurred in 1975.  The
PDP-10 was a 36-bit mainframe computer. It represented dates using a 12-bit
format designed for the tiny PDP-8. With 12 bits, one can distinguish
$2^{12} = 4096$ days or 11 years.

The most common industry format for dates uses six characters: two for the
year, two for the month and two for the day. The most common "solution" to the
year 2000 crisis is to add two further characters, thereby altering file sizes.
Others have noticed that the existing six characters consist of 48 bits,
already sufficient to represent all dates over the projected lifetime of the
universe: $2^{48}$ = $2.8 * 1014$ days = $7.7 * 1011$ years!

Mathematicians think in terms of unbounded ranges, but the representation we
choose for the computer usually imposes hard limits. A good programming
language like OCaml lets one easily change the representation used in the
program.  But if files in the old representation exist all over the place,
there will still be conversion problems. The need for compatibility with older
systems causes problems across the computer industry.

#### Example II: Floating Point Numbers

Computers have integers like `1066` and floats like $1.066 x 10^3$.
A floating-point number is represented by two integers.
The concept of _data type_ involves:
* how a value is represented inside the computer
* the suite of operations given to programmers
* valid and invalid (or exceptional) results, such as “infinity”
Computer arithmetic can yield _incorrect answers_!

In science, numbers written with finite precision and a decimal exponent are
said to be in _standard form_. The computational equivalent is the _floating
point number_. These are familiar to anybody who has used a scientific
calculator.  Internally, a float consists of two integers.

Because of its finite precision, floating-point computations are potentially
inaccurate. To see an example, use your nearest electronic calculator to
compute $(2^{1/10000})10000$. I get $1.99999959$! With certain computations,
the errors spiral out of control. Many programming languages fail to check
whether even integer computations fall within the allowed range: you can add
two positive integers and get a negative one!

Most computers give us a choice of precisions. In 32-bit precision, integers
typically range from $2^{31} − 1$ (namely $2,147,483,647$) to $−2^{31}$; reals
are accurate to about six decimal places and can get as large as 1035 or so.
For reals, 64-bit precision is often preferred. Early languages like Fortran
required variables to be declared as `INTEGER`, `REAL` or `COMPLEX` and barred
programmers from mixing numbers in a computation. Nowadays, programs handle
many different kinds of data, including text and symbols. The concept of a
_data type_ can ensure that different types of data are not combined in a
senseless way.

Inside the computer, all data are stored as bits. In most programming
languages, the compiler uses types to generate correct machine code, and types
are not stored during program execution. In this course, we focus almost
entirely on programming in a high-level language: OCaml.

#### Goals of Programming

- to describe a computation so that it can be done _mechanically_:
  * Expressions compute values.
  * Commands cause effects.
- to do so efficiently and **correctly**, giving the right answers quickly
- to allow easy modification as needs change
  * Through an orderly _structure_ based on abstraction principles
  * Such as modules or (Java) classes

Programming _in-the-small_ concerns the writing of code to do simple, clearly
defined tasks. Programs provide expressions for describing mathematical
formulae and so forth. (This was the original contribution of FORTRAN, the
FORmula TRANslator.) Commands describe how control should flow from one part of
the program to the next.

As we code layer upon layer, we eventually find ourselves programming
_in-the-large_: joining large modules to solve some messy task. Programming
languages have used various mechanisms to allow one part of the program to
provide interfaces to other parts. Modules encapsulate a body of code, allowing
outside access only through a programmer-defined interface. _Abstract Data
Types_ are a simpler version of this concept, which implement a single concept
such as dates or floating-point numbers.

_Object-oriented programming_ is the most complicated approach to modularity.
_Classes_ define concepts, and they can be built upon other classes. Operations
can be defined that work in appropriately specialized ways on a family of
related classes. _Objects_ are instances of classes and hold the data that is
being manipulated.

This course does not cover OCaml's sophisticated module system, which can do
many of the same things as classes. You will learn all about objects when you
study Java.

### Why Program in ML?

* Why Program in ML?
* It is interactive.
* It has a flexible notion of _data type_.
* It hides the underlying hardware: _no crashes_.
* Programs can easily be understood mathematically.
* It distinguishes naming something from _updating memory_.
* It manages storage for us.

Standard ML is the outcome of years of research into
programming languages. It is unique, defined using a mathematical formalism (an
operational semantics) that is both precise and comprehensible. Several
supported compilers are available, and thanks to the formal definition, there
are remarkably few incompatibilities among them. _(TODO edit)_

Because of its connection to mathematics, ML programs can be designed and
understood without thinking in detail about how the computer will run them.
Although a program can abort, it cannot crash: it remains under the control of
the OCaml system. It still achieves respectable efficiency and provides
lower-level primitives for those who need them. Most other languages allow
direct access to the underlying machine and even try to execute illegal
operations, causing crashes.

The only way to learn programming is by writing and running programs. This
web notebook provides an interactive environment where you can modify
the example fragments and see the results for yourself.  You should also
consider installing OCaml on your own computer so that you try more
advanced programs locally.

### A first session with OCaml

In [2]:
let pi = 3.1415926535

val pi : float = 3.1415926535


The first line of this simple session is a _value declaration_. It makes the
name `pi` stand for the floating point number `3.14159`. (Such names are called
_identifiers_.)  OCaml echoes the name (`pi`) and type (`float`) of the
declared identifier.


In [3]:
pi *. 1.5 *. 1.5

- : float = 7.068583470375


The second line computes the area of the circle with radius `1.5` using the
formula $A = \pi r^2$. We use `pi` as an abbreviation for `3.14159`.
Multiplication is expressed using `*.`, which is called an _infix operator_
because it is written between its two operands.

OCaml replies with the computed value (about `7.07`) and its type (again `float`).


In [4]:
let area r = pi *. r *. r

val area : float -> float = <fun>


To work abstractly, we should provide the service "compute the area of a
circle," so that we no longer need to remember the formula. This sort of
encapsulated computation is called a _function_. The third line declares the
function `area`. Given any floating point number `r`, it returns another
floating point number computed using the `area` formula; note that the function
has type `float -> float`.


In [5]:
area 2.0

- : float = 12.566370614


The fourth line calls the function `area` supplying `2.0` as the argument. A
circle of radius `2` has an area of about `12.6`. Note that brackets around a
function argument are not necessary.

The function uses `pi` to stand for `3.14159`. Unlike what you may have seen in
other programming languages, `pi` cannot be "assigned to" or otherwise updated.
Its meaning within `area` will persist even if we issue a new `let` declaration
for `pi` afterwards.


# Remainder of Content

### Exercise 1.2
Using the date representation of the previous exercise, code OCaml functions to:
- compare two years
- add/subtract some given number of years from another year.

(You may need to look ahead to the next lecture for OCaml’s comparison operators.)


In [6]:
let x = "answer"

val x : string = "answer"


In [7]:
let toSeconds minutes seconds =
  seconds +. (60.0 *. minutes)

val toSeconds : float -> float -> float = <fun>


In [8]:
let fromSeconds seconds =
  (seconds /. 60.0), (mod_float seconds 60.0)

val fromSeconds : float -> float * float = <fun>


In [9]:
let rec npower x n =
  if n = 0 then 1.0
  else x *. npower x (n-1)

val npower : float -> int -> float = <fun>


### Type constraints

In [10]:
let square x = x *. x;

val square : float -> float = <fun>


In [11]:
let square x : float = x *. x

val square : float -> float = <fun>


In [12]:
let square (x:float) = x *. x

val square : float -> float = <fun>


A boolean-valued function to test whether a number is even:


In [13]:
let even n =
n mod 2 = 0

val even : int -> bool = <fun>


raising to an integer power--fast version


In [14]:
let rec power x n : float =
  if n = 1 then x
  else if even n then      power (x *. x) (n / 2)
                 else x *. power (x *. x) (n / 2)

val power : float -> int -> float = <fun>


the sum of the first n integers


In [15]:
let rec nsum n =
  if n = 0 then 0
           else n + nsum (n-1)

val nsum : int -> int = <fun>


In [16]:
let rec summing n total =
  if n = 0 then total
           else summing (n-1) (n + total)

val summing : int -> int -> int = <fun>


with an exponential runtime


In [17]:
let rec stupidSum n =
  if n = 0 then 0
           else n + (stupidSum (n-1) + stupidSum (n-1)) / 2

val stupidSum : int -> int = <fun>
